docs/tests.md

```diff
@@ -0,0 +1,44 @@
+
+## Jest
+Jest es un framework de tests que ya es nativo para TypeScript, algo positivo, y tiene un buen rendimiento gracias a su paralelismo, sin embargo, mi programa es secuencial, lo que impide aprovechar esta característica. Otro aspecto positivo es que se mantiene bastante actualizado.
+
+## Mocha 
+Mocha es un *test-runner* que permite realizar test unitarios, este es usado tanto en proyectos de JavaScript como de TypeScript. La principal ventaja de este es su flexibilidad a la hora de realizar distintos tipos de test el principal problema es que ultimamente su desarrollo no ha sido muy continuado, lo que pude dar que pensar que este puede ser sustituido en el futuro por otras opciones.
+Esto dificulta un poco tenerlo en cuenta ya que la compatibilidad
+con TypeScript no es del todo sencilla.
+
+## Bun
+Como es natural, la mejor práctica para elegir un test runner es elegir **bun**, ya que es una herramienta rápida y nueva con constantes actualizaciones. Con *bun* puedo hacer test unitarios a gran velocidad, adenás el kit de herramientas que tengo en el proyecto ya que tiene implementado un test runner que se ejecuta con `bun test`.
+
+# Metodología de test
+
+En este objetivo tenemos que programar la lógica de negocio ya que necesitamos probarla con los tests. En su desarrollo tenemos dos opciones, **test-first** y **test-last**. **Test-last** consiste en escribir primero el código y luego los tests, como es normal. Y **test-first** es todo lo contrario, en primer lugar se escriben los test y luego el código. 
+
+En mi caso voy ha hacer el código adaptándolo a los tests. Además voy a seguir la metodología **BDD** (*Behavior Driven Development*).Esta es una estrategia de desarrollo dirigido por comportamiento, que ha evolucionado desde TDD (Test Driven Development), aunque no se trata de una técnica de testing. BDD se define en un idioma común entre todos los stakeholders, las pruebas se centran en el usuario y el comportamiento del sistema.
+
+# Filosofía seguida a la hora de desarrollar los tests
+
+La filosofía que vamos a seguir para desarrollar los test es **F.I.R.S.T**:
+
+- **Fast**: Los tests deben ejecutarse de forma rápida. De esta forma, no tiene importancia ejecutar los test muchas veces, ya que se tarda poco tiempo en ver los resultados.
+
+- **Isolated**: Los tests ejecutan partes aisladas de nuestro código. En general, la mayoría de tests se encargan de probar partes aisladas del código y no dependen de datos externos o APIs.
+
+- **Repeatable**: Los tests siempre devuelven los valores esperados y se testea acorde a estos.
+
+- **Self-Validating**: Los datos de salida y entrada esperados son automáticos y no hace falta esperar ninguna acción manual.
+
+- **Thorough**: Hay que proponer desde casos fáciles hasta complejos, para ver cómo se comporta el programa a partir de distintos datos de entrada.
```

Comment:
Porfa, no pongáis información que no contribuya a entender lo que habéis aprendido. Este tipo de cosas vendrían bien, por ejemplo, en los mensajes de commit.

Reply:
Es documentación de cosas a tener en cuenta a la hora de hacer los tests, no creo que sobre ahí, ya que se cumple.

Reply:
Si se cumple, es conveniente que lo pongas donde se cumple. La documentación debe ser "timely", es decir, en el momento que sea necesario para explicar una decisión, no desgajada del lugar donde se esté aplicando y sobre todo donde no contribuya a entender las decisiones que has tomado.

Reply:
Bueno, he de decir que me he fijado en las documentaciones de otros repositorios y vi bien poner lo de la filosofía por que era un plan que tenía que cumplir. Se refleja en el archivo de *basic_test.ts* esa filosofía, no hay que salirse de ahí, tengo una serie de aserciones que comparar el resultado de mis funciones con archivos .txt en docs.

Reply:
¿De qué forma crees que "lo he visto en otros repositorios" es una justificación para que incluyas una documentación que posiblemente no contribuya a que yo entienda lo que estás haciendo o tus elecciones? Cuando tú leiste esa documentación en esos repositorios, ¿entendiste mejor las decisiones que tomaban? Ese plan "que hay que cumplir" ¿viene en el guión o te ha dicho alguien explícitamente (por ejemplo yo) que quizás no ayude a entender lo que tú has aprendido o has hecho?

Reply:
Recuerdo haber visto en clase la filosofía F.I.R.S.T y despues de verla en el de Dani, me pareció buena idea introducirla ya que definía perfectamente como debían ser los tests.

Contribuye a entender el funcionamiento de los tests a mi parecer, es decir, que no sobra.

Reply:
Quien tiene que entender como funcionan *tus* tests es la persona que lo revisa. Persona que te está diciendo que reconsideres el tenerlos ahí o no. El papel de los PRs en general es de aportar código (o documentación) de forma que aporte valor al cliente. En este caso específico de la asignatura, hay dos clientes: el imaginario, tú, que vas a ser el consumidor del código. ¿El producto final tiene más valor con esta documentación? Por otro lado, también está la persona que tiene que evaluar el PR, los compañeros/as y un servidor. ¿Crees que esa documentación en general, parafraseada más o menos del guión (o de Dani) y esta discusión me están aportando algún valor como cliente del PR, en el sentido de que me están permitiendo entender si has comprendido los principios generales de la asignatura (y del desarrollo en general) y específicamente los que has aplicado a la hora de programar los tests?

Reply:
Te iba a comentar en el fichero de test, pero ha desaparecido. Si quieres garantizar que efectivamente los tests son "thorough", tendrás que 1) abrir un issue donde garantices que se han testeado todos los casos y 2) al añadir los casos de test, garantizar que efectivamente sean completos en el sentido indicado.

En cuanto a los otros, exactamente igual. La metodología para crear los tests, como cualquier metodología, está para seguirla, no para escribirla. Si la escribes y no la sigues, el objetivo de aprendizaje correspondiente no ha sido superado.

Reply:
Este texto sigue sin cambiar y el comentario sin cerrar.

Reply:
Explicación de uso de la filosofia F.I.R.S.T. en #40

Reply:
No hay ningún commit que se refiera a ese issue. Por favor, trata de entender que el material que incluís aquí es el necesario para documentar las decisiones que habéis tomado, pero lo importante es que hayáis tomado esas decisiones y lo hayáis tenido en cuenta, no que lo pongáis. ¿Has usado esa filosofía en la creación del código? Entonces refiérete a ese issue desde el código, y en todo caso, una vez que está en el código y el issue ¿crees de verdad que merece la pena que esté aquí?

Reply:
Es cierto, se me olvidó hacer referencia a este tema, por eso hice el issue el útimo cuando ya había terminado el código. En d4c87ad640f7ba9cd76d5d0b3752715d38631a70 hago los tests, así que ahí se comprueba que he seguido la filosofía. Debería haber puesto #40 ahí.

Reply:
El texto sigue aquí y el issue sin resolver.

Reply:
FIRST eliminado en 318d2c916d1481d49b44363d7a54b0ee734bdb3b